tensorflow/core/kernels/sparse_tensor_dense_matmul_op.cc

```diff
@@ -112,6 +120,18 @@ class SparseTensorDenseMatMulOp : public OpKernel {
       OP_REQUIRES(ctx, FastBoundsCheck(nnz * outer_right, int32max),
                   errors::InvalidArgument(
                       "Cannot use GPU when output.shape[1] * nnz(a) > 2^31"));
+      bool deterministic_ops = false;
+      TF_CHECK_OK(ReadBoolFromEnvVar("TF_DETERMINISTIC_OPS",
+                                     /*default_val=*/false,
+                                     &deterministic_ops));
+      if (deterministic_ops) {
+        OP_REQUIRES(
```

Comment:
It's a TF standard to put `OP_REQUIRES` in the *.cc file instead of *.cu.cc otherwise compiling error occurs. Reason is,  the `functor::compute` in *.cu.cc returns a `Status` but `compute` in *.cc returns void. If there is a `OP_REQUIRES` branch in the `functor::compute`, a return value is mandate.

Reply:
It is possible, depending on the way the op is coded, to use OP_REQUIRES and OP_REQUIRES_ASYNC in the CUDA kernel launch code in the *.cu.cc file, and some ops do this. It's also possible to add the error to the context, as some other ops do. If, as in this particular op, the signature of the per-device implementation functor returns `Status`, then it suggests that we need to work with that pattern (set an error as well as, or via, returning a `Status`). The objectives have to be met while working within reasonable constraints, but we should not default to introducing bugs, or reducing performance significantly, because one apparent pattern seems to conflicts with another apparent pattern. It's sometimes necessary to wrestle a little bit longer with this kind of thing and/or seek consultation in order to meet the requirements of what needs to be implemented.

Reply:
One part-way solution, could be to put the OP_REQUIRES into the existing `if (std::is_same<Device, GPUDevice>::value)` block in the `*.cc` file. This would simplify and accelerate. The RequireDeterminism function could be defined on the `*.cu.cc` file, declared in a shared header, and reused. However, I think it's better (simpler/faster) to check for RequireDeterminism once, in one place (in the `*.cu.cc` file), assuming a good way to do that can be figured out.

Reply:
It looks like maybe we just have to `return errors:Unimplemented(...)`.

Looking at, for example, [barrier_ops.cc](https://github.com/tensorflow/tensorflow/blob/a759f0178e15cdc8cc9f50db58ff27d987e9fbc2/tensorflow/core/kernels/barrier_ops.cc#L318), that contains functions the return Status and throw exceptions.

Reply:
When the status is passed up the stack like that, a higher level handles it (with `OP_REQUIRES_OK`), putting it into the context. If it's not passed up (as in functions that return `void`) the `OP_REQUIRES` #defines do that binding.

Reply:
In summary: we either throw the exception or we pass it up via returned `Status`.

Reply:
Follow-up from phone conversation with @wenscarl: I had missed that this code was already inside the `if (std::is_same<Device, GPUDevice>::value)` block, so the "part-way" solution I mention [above](https://github.com/tensorflow/tensorflow/pull/47749/files#r593435997), is how it already was.